src/daemon/daemon.cpp

```diff
@@ -681,7 +682,7 @@ try // clang-format on
             reply.set_metrics_pending(true);
             server->Write(reply);
 
-            return grpc::Status::OK;
+            return status_promise->set_value(grpc::Status::OK);
```

Comment:
Rather than changing all the code to replace every status return by the promise setting bit, I would prefer to see that encapsulated in a single place. For instance, a template that received a callable, forwarded the params to it, and set the return value in the promise. The signals would then be connected with instantiations of this template for launch, start, stop, etc. Changes would then be greatly simplified, the history of the code would be clearer, easier to review, and dozens of repetitions eliminated.

Reply:
Ok, sure, I will look into that.

Reply:
After thinking about this, I'm not really sure what you're wanting to achieve here.  Could you please provide a small example?

Reply:
I may be missing something, but my thought was: *since the promise is always being set to what was previously returned*, instead of changing each return to a promise setting call, we could have an intermediate thin layer to do that. This layer (let's call it `promise_fulfiller`) would sit between the rpc and `daemon::start`, `stop`, `launch`, etc. Those operations would still return the status (as until now), which the `promise_fulfiller` would use to set the promise. 

The `promise_fulfiller`  layer would probably be a template, receiving a method to call (one of `start`, `stop`, etc.), along with the corresponding parameters. The slots would not be `start`, `stop`, etc. directly, but an instantiation of `promise_fulfiller` with one of those methods.

Something along these lines (please treat this as pseudocode sketching):

```cpp
template<typename Operation, typename Request, typename Reply>
void promise_fulfiller(Operation op, Request request, Reply reply, std::promise promise)
{
  promise.set(op(request, reply));
}
...
QObject::connect(&rpc, &mp::DaemonRpc::on_launch, std::bind(promise_fulfiller, &mp::Daemon::launch));
```

Reply:
I see, interesting...I'll see about changing that.  Thanks!

Reply:
Hmm, well, now that I think about it a bit more, I don't think this will work.  The operation itself doesn't necessarily set the value of the promise, but can be done in `mp::Daemon::finish_async_operation()` when there is an actual asynchronous operation.

Reply:
I agree with @townsend2010 here, it makes sense to permit a promise to be passed elsewhere, as it may take an operation on a separate thread to complete before the promise can be fulfilled, and we want the main loop unblocked asap.